File: ios/WebyarNative/Sources/Features/Inbox/InboxViewModel.swift
import SwiftUI
import Observation

/// What a list screen can be showing at any moment.
///
/// Modelling this as one value rather than a pile of booleans is what makes it
/// impossible to render a spinner and an empty state at the same time.
enum LoadState<Value>: Sendable where Value: Sendable {
    case loading
    case loaded(Value)
    case failed(APIError)

    var value: Value? {
        if case .loaded(let value) = self { return value }
        return nil
    }

    /// True once the rows on screen are real content rather than placeholders.
    var isLoaded: Bool {
        if case .loaded = self { return true }
        return false
    }
}

/// How the last attempt to bring what is on screen up to date went.
///
/// Apart from `LoadState` on purpose. `LoadState` says what content there is
/// to show; this says how fresh it is. With a copy saved on the phone a
/// failed read no longer means "nothing to show": the rows stay, and this
/// says they are the saved ones — which is the difference between an inbox
/// that disappears in a lift and one that says "offline" above what it had.
enum SyncStatus: Equatable, Sendable {
    /// Up to date as far as anyone knows.
    case idle
    /// The last read could not reach the server; the saved copy is on show.
    case offline
    /// The server answered with a failure; what was shown before is kept.
    case failed(APIError)

    var isOffline: Bool { self == .offline }
}

@MainActor
@Observable
final class InboxViewModel {

    private(set) var state: LoadState<[Conversation]> = .loading
    /// How fresh `state` is. Only meaningful while `state` is `.loaded`.
    private(set) var syncStatus: SyncStatus = .idle
    /// The counters behind each queue, so a filter can say how much is in it
    /// before the operator taps into it.
    private(set) var counts: InboxCounts?
    /// Device and location per conversation, keyed by conversation id.
    ///
    /// Decorative, exactly as in the web: the list renders without it, so a
    /// failure here never becomes an error state.
    private(set) var visitors: [String: VisitorProfile] = [:]
    var filter: InboxFilter = .open
    /// A channel inbox laid over the queue — Telegram, Bale, and the rest.
    ///
    /// Not a queue: the conversations endpoint has no channel parameter, so
    /// this narrows what came back, exactly as `?channel=` does on the web.
    var channel: ChannelInbox?
    /// Which channels this workspace has installed and can still use.
    private(set) var channels: [ChannelInbox] = []
    var searchText = ""
    /// The filter sheet's three fields.
    var fieldFilter = InboxFieldFilter()

    /// The workspace and queue `state` belongs to. A screen asking for
    /// another workspace's rows is told `.loading` until they are its own —
    /// not one frame of the previous workspace's list.
    private(set) var stateWorkspaceID: String?
    @ObservationIgnored private var stateFilter: InboxFilter?

    /// Set by the screen: the list is what the operator is looking at (its
    /// tab, nothing pushed over it). A list nobody is looking at is not read
    /// again for every event; it is read once when it is looked at again.
    @ObservationIgnored var isOnScreen = true {
        didSet { if isOnScreen, !oldValue, isStale { scheduleRefresh(after: 0) } }
    }

    @ObservationIgnored private let api: any WebyarAPI
    @ObservationIgnored private let sync: SyncCoordinator
    @ObservationIgnored private var loadTask: Task<Void, Never>?
    @ObservationIgnored private var refreshSoon: Task<Void, Never>?
    @ObservationIgnored private var generation = 0
    @ObservationIgnored private weak var appState: AppState?
    /// Something changed on the server since the list was last read.
    @ObservationIgnored private var isStale = false

    init(api: any WebyarAPI = Backend.current, sync: SyncCoordinator = .shared) {
        self.api = api
        self.sync = sync
    }

    /// Keeps the selected queue inside the set the plan actually grants.
    ///
    /// A plan can change under the app — switching workspace is the ordinary
    /// way — and leaving the selection on a queue that no longer exists would
    /// show an empty list with no visible reason.
    func reconcileFilter(with available: [InboxFilter]) {
        guard !available.isEmpty, !available.contains(filter) else { return }
        filter = available[0]
    }

    /// What the screen for `workspaceID` may show.
    func content(for workspaceID: String?) -> LoadState<[Conversation]> {
        guard stateWorkspaceID == workspaceID else { return .loading }
        return state
    }

    /// One conversation out of the loaded list, by id.
    ///
    /// The whole list, not `visible`: a notification names a conversation
    /// without knowing which queue the operator happens to be looking at, and
    /// refusing to open a thread because it is filtered out of the current
    /// view would be the app disagreeing with its own banner.
    func conversation(id: String) -> Conversation? {
        state.value?.first { $0.id == id }
    }

    /// Conversations after the queue, the search box and the filter sheet,
    /// sorted newest first — for `workspaceID` only.
    ///
    /// All three are applied here rather than on the server because
    /// `GET /api/conversations` has no text search — the web filters its own
    /// list the same way, so the two surfaces agree on what a match is.
    func visible(in workspaceID: String?) -> [Conversation] {
        guard stateWorkspaceID == workspaceID, let all = state.value else { return [] }
        let query = searchText.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()

        let searched = query.isEmpty ? all : all.filter { conversation in
            let haystack = [
                conversation.contact?.name,
                conversation.contact?.email,
                conversation.contact?.visitorCode,
                conversation.subject,
                conversation.lastMessage?.body,
            ]
            return haystack.contains { $0?.lowercased().contains(query) == true }
        }

        let matched = fieldFilter.isEmpty ? searched : searched.filter(fieldFilter.matches)
        let onChannel = channel.map { inbox in
            matched.filter { $0.channelKey == inbox.key }
        } ?? matched

        return onChannel
            // Belt and braces: a row of another workspace is never drawn.
            .filter { $0.workspaceId == workspaceID }
            .sorted { lhs, rhs in
                // A thread with no activity at all sorts last rather than
                // crashing into the top on a nil date.
                (lhs.lastActivity ?? .distantPast) > (rhs.lastActivity ?? .distantPast)
            }
    }

    /// The channels the switcher can offer. Decorative in the same sense the
    /// visitor lookups are: a failure leaves the menu with its queues and
    /// nothing else, which is the state the app shipped in.
    func loadChannels(workspaceID: String?) async {
        guard let workspaceID else { return }
        channels = (try? await api.channelInboxes(workspaceID: workspaceID)) ?? []
        // A channel that has just left the plan must not stay selected with
        // nothing behind it.
        if let channel, !channels.contains(channel) { self.channel = nil }
    }

    /// Opens one inbox from the switcher: either a queue or a channel, never
    /// both at once. They read as siblings in the menu, so they behave as
    /// siblings here.
    func open(_ filter: InboxFilter) {
        self.filter = filter
        channel = nil
    }

    func open(_ channel: ChannelInbox) {
        self.channel = channel
        // A channel inbox shows what is open on it; the other queues are
        // reachable from the same menu.
        filter = .open
    }

    /// Shows the list for this workspace and queue: this session's copy or
    /// the one saved on the phone at once, then the server's.
    func load(workspaceID: String?, appState: AppState) {
        self.appState = appState
        guard let workspaceID else {
            stateWorkspaceID = nil
            stateFilter = nil
            state = .loaded([])
            return
        }

        // A fast filter or workspace switch must not let an older, slower
        // response land on top of a newer one.
        generation += 1
        let gen = generation
        loadTask?.cancel()
        let filter = self.filter

        if workspaceID != stateWorkspaceID {
            counts = nil
            visitors = [:]
            syncStatus = .idle
        }
        if workspaceID != stateWorkspaceID || filter != stateFilter {
            // Synchronously, so the first frame for the new list is already
            // the right one: this session's copy, or a skeleton.
            if let known = sync.lists(for: workspaceID)?.cachedInMemory(filter) {
                show(known, workspaceID: workspaceID, filter: filter)
            } else {
                stateWorkspaceID = workspaceID
                stateFilter = filter
                state = .loading
            }
        }

        loadTask = Task {
            // The copy saved on the phone while the server is asked.
            if !state.isLoaded, let lists = sync.lists(for: workspaceID) {
                if let saved = await lists.cached(filter), gen == generation, !state.isLoaded {
                    show(saved, workspaceID: workspaceID, filter: filter)
                }
                if counts == nil, let saved = await sync.store(for: workspaceID)?.counts(), gen == generation, counts == nil {
                    counts = saved
                }
            }
            await fetch(workspaceID: workspaceID, filter: filter, generation: gen)
        }
    }

    /// Pull-to-refresh. Unlike `load` it keeps the current rows on screen
    /// while the request is in flight, so the list does not blank out under
    /// the finger.
    func refresh(workspaceID: String?, appState: AppState) async {
        self.appState = appState
        guard let workspaceID else { return }
        sync.lists(for: workspaceID)?.invalidate()
        await fetch(workspaceID: workspaceID, filter: filter, generation: generation)
    }

    private func show(_ list: [Conversation], workspaceID: String, filter: InboxFilter) {
        let own = list.filter { $0.workspaceId == workspaceID }
        stateWorkspaceID = workspaceID
        stateFilter = filter
        // Unchanged rows are not re-published: a poll that finds nothing new
        // redraws nothing.
        if case .loaded(let current) = state, current == own { return }
        state = .loaded(own)
    }

    private func fetch(workspaceID: String, filter: InboxFilter, generation gen: Int) async {
        isStale = false
        async let counters = try? await api.inboxCounts(workspaceID: workspaceID, scope: "mine")
        do {
            let conversations: [Conversation]
            if let lists = sync.lists(for: workspaceID) {
                conversations = try await lists.fetch(filter)
            } else {
                conversations = try await api.conversations(workspaceID: workspaceID, filter: filter)
            }
            // The counters are a nice-to-have on top of the list, so a
            // failure there must not empty the inbox.
            let fetchedCounts = await counters
            guard gen == generation, !Task.isCancelled else { return }
            show(conversations, workspaceID: workspaceID, filter: filter)
            syncStatus = .idle
            if let fetchedCounts {
                if counts != fetchedCounts { counts = fetchedCounts }
                await sync.store(for: workspaceID)?.saveCounts(fetchedCounts)
            }
            await loadVisitors(for: conversations, workspaceID: workspaceID)
        } catch APIError.unauthorized {
            await appState?.handleUnauthorized()
        } catch {
            guard gen == generation, !Task.isCancelled else { return }
            let apiError = (error as? APIError) ?? .transport
            if state.isLoaded, stateWorkspaceID == workspaceID {
                // The rows stay; the screen says they may not be current.
                syncStatus = apiError == .transport ? .offline : .failed(apiError)
            } else {
                state = .failed(apiError)
            }
        }
    }

    // MARK: - Staying fresh

    /// Listens for as long as the screen exists: realtime events, pushes and
    /// the return to the foreground read the list again (one read for a
    /// burst); without realtime, the list is polled while it is on screen.
    func listen(workspaceID: String?) async {
        guard let workspaceID else { return }
        await withTaskGroup(of: Void.self) { group in
            group.addTask { await self.consumeEvents(workspaceID: workspaceID) }
            group.addTask { await self.poll(workspaceID: workspaceID) }
        }
    }

    private func consumeEvents(workspaceID: String) async {
        for await event in sync.events() {
            guard stateWorkspaceID == workspaceID else { continue }
            switch event {
            case .message, .conversationChanged, .push, .resync, .reconcile:
                isStale = true
                if isOnScreen { scheduleRefresh(after: CachePolicy.listEventDebounce) }
            }
        }
    }

    private func poll(workspaceID: String) async {
        while !Task.isCancelled {
            let interval = sync.realtimeConnected ? CachePolicy.listSafetyInterval : CachePolicy.listPollInterval
            try? await Task.sleep(nanoseconds: UInt64(interval * 1_000_000_000))
            guard !Task.isCancelled else { return }
            guard sync.isForeground, isOnScreen, stateWorkspaceID == workspaceID else { continue }
            sync.lists(for: workspaceID)?.invalidate()
            await fetch(workspaceID: workspaceID, filter: filter, generation: generation)
        }
    }

    private func scheduleRefresh(after delay: TimeInterval) {
        refreshSoon?.cancel()
        refreshSoon = Task {
            if delay > 0 { try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000)) }
            guard !Task.isCancelled, let workspaceID = stateWorkspaceID else { return }
            // Something changed on the server: an answer from before it is
            // not shared, however recent.
            sync.lists(for: workspaceID)?.invalidate()
            await fetch(workspaceID: workspaceID, filter: filter, generation: generation)
        }
    }

    /// One batched lookup per page of conversations, never one per row — and
    /// only for the rows not already known from a moment ago.
    private func loadVisitors(for conversations: [Conversation], workspaceID: String) async {
        guard !conversations.isEmpty else { return }
        let cache = VisitorIntelCache.shared
        let ids = conversations.map(\.id)
        var known: [String: VisitorProfile] = [:]
        for id in ids { if let profile = cache.profile(for: id) { known[id] = profile } }
        let missing = cache.missing(ids)
        if !missing.isEmpty,
           let fetched = try? await api.visitorIntel(workspaceID: workspaceID, conversationIDs: missing) {
            cache.store(fetched, workspaceID: workspaceID)
            known.merge(fetched) { _, new in new }
        }
        guard !known.isEmpty, stateWorkspaceID == workspaceID else { return }
        // Merged rather than replaced: switching queue re-uses what is already
        // known about a thread instead of blanking its avatar for a moment.
        var merged = visitors
        merged.merge(known) { _, new in new }
        if merged.count != visitors.count || known.contains(where: { visitors[$0.key] == nil }) {
            visitors = merged
        }
    }

    func visitor(for conversation: Conversation) -> VisitorProfile? {
        visitors[conversation.id]
    }

    /// Takes ownership of a thread.
    ///
    /// Unlike resolving, this does not remove the row — the conversation stays
    /// in the same queue, it just becomes yours — so the list is reloaded
    /// rather than mutated, and the server decides what "yours" now means.
    func claim(_ conversation: Conversation, workspaceID: String?, appState: AppState) async {
        do {
            try await api.claim(conversationID: conversation.id, workspaceID: conversation.workspaceId)
            Haptics.success()
            await refresh(workspaceID: workspaceID, appState: appState)
        } catch APIError.unauthorized {
            await appState.handleUnauthorized()
        } catch {
            // Nothing changed server-side, so nothing changes here either.
        }
    }

    func setStatus(_ status: ConversationStatus, for conversation: Conversation, appState: AppState) async {
        // Optimistic: the row animates away immediately, because waiting for a
        // round trip to acknowledge a swipe feels broken.
        let previous = state.value
        if var list = state.value, let index = list.firstIndex(where: { $0.id == conversation.id }) {
            list.remove(at: index)
            state = .loaded(list)
        }

        do {
            try await api.setStatus(status, conversationID: conversation.id, workspaceID: conversation.workspaceId)
            // Every list held for this workspace forgets the row's old place;
            // the next read (the event this change publishes) confirms it.
            sync.lists(for: conversation.workspaceId)?.remove(conversation.id)
        } catch APIError.unauthorized {
            await appState.handleUnauthorized()
        } catch {
            // Put it back — the server never agreed to the change.
            if let previous, stateWorkspaceID == conversation.workspaceId { state = .loaded(previous) }
        }
    }
}

/// The three things an operator actually looks a conversation up by.
///
/// Separate from the search box on purpose: search is one word against
/// everything, this is a specific word against a specific field, and the two
/// compose — a subject filter narrows what the search box already found.
struct InboxFieldFilter: Equatable, Sendable {
    var name = ""
    var email = ""
    var subject = ""

    var isEmpty: Bool {
        [name, email, subject].allSatisfy {
            $0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        }
    }

    var activeCount: Int {
        [name, email, subject].count {
            !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        }
    }

    func matches(_ conversation: Conversation) -> Bool {
        contains(conversation.contact?.name, name)
            && contains(conversation.contact?.email, email)
            && contains(conversation.subject, subject)
    }

    private func contains(_ haystack: String?, _ needle: String) -> Bool {
        let wanted = needle.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard !wanted.isEmpty else { return true }
        return haystack?.lowercased().contains(wanted) == true
    }
}

File: android/app/src/main/kotlin/com/webyar/operator/feature/inbox/InboxViewModel.kt
package com.webyar.operator.feature.inbox

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.webyar.operator.core.cache.CacheScope
import com.webyar.operator.core.model.ChannelInbox
import com.webyar.operator.core.model.Conversation
import com.webyar.operator.core.model.Entitlements
import com.webyar.operator.core.model.InboxCounts
import com.webyar.operator.core.model.InboxFilter
import com.webyar.operator.core.model.VisitorProfile
import com.webyar.operator.core.model.WorkspaceAccess
import com.webyar.operator.core.model.channelKey
import com.webyar.operator.core.net.WebyarApi
import com.webyar.operator.core.sync.SyncGraph
import com.webyar.operator.core.sync.isOffline
import com.webyar.operator.i18n.StrAndroid
import com.webyar.operator.i18n.Format
import com.webyar.operator.i18n.Language
import com.webyar.operator.i18n.displayText
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * The inbox: which queue, which channel, what is in it, and how much.
 *
 * **Local-first.** What the screen shows is the cache ([SyncGraph]'s Room
 * flow for this scope and queue), from the first frame of a cold start —
 * the network only ever writes the cache, and the cache's flow carries the
 * change here. A failed refresh therefore never costs the operator the list
 * they were reading: with rows on screen it becomes [syncProblem], a line
 * above the list, and only a queue that has never once been read can end in
 * [InboxState.Failed].
 *
 * Search is done HERE rather than by the server, because the conversations
 * endpoint has no query parameter — the console narrows its own list the same
 * way. That is fine for a phone, where the page is fifty rows; it would not be
 * fine for a desktop showing a thousand, which is why this is stated rather
 * than left to look like an oversight.
 */
class InboxViewModel(
    private val api: WebyarApi,
    /** Last-but-one so `InboxViewModel(api) { language }` still reads as it always did. */
    private val sync: SyncGraph = SyncGraph.inMemory(api),
    private val language: () -> Language,
) : ViewModel() {

    private val _state = MutableStateFlow<InboxState>(InboxState.Loading)
    val state: StateFlow<InboxState> = _state.asStateFlow()

    private val _filter = MutableStateFlow(InboxFilter.OPEN)
    val filter: StateFlow<InboxFilter> = _filter.asStateFlow()

    private val _counts = MutableStateFlow(InboxCounts())
    val counts: StateFlow<InboxCounts> = _counts.asStateFlow()

    private val _channels = MutableStateFlow<List<ChannelInbox>>(emptyList())
    val channels: StateFlow<List<ChannelInbox>> = _channels.asStateFlow()

    /** Null means every channel, which is the ordinary case. */
    private val _channel = MutableStateFlow<String?>(null)
    val channel: StateFlow<String?> = _channel.asStateFlow()

    private val _query = MutableStateFlow("")
    val query: StateFlow<String> = _query.asStateFlow()

    private val _intel = MutableStateFlow<Map<String, VisitorProfile>>(emptyMap())
    val intel: StateFlow<Map<String, VisitorProfile>> = _intel.asStateFlow()

    private val _refreshing = MutableStateFlow(false)
    val refreshing: StateFlow<Boolean> = _refreshing.asStateFlow()

    /**
     * Why the list on screen may be out of date — offline, the server
     * refusing — or null when the last read worked. Shown above the rows,
     * never instead of them.
     */
    private val _syncProblem = MutableStateFlow<String?>(null)
    val syncProblem: StateFlow<String?> = _syncProblem.asStateFlow()

    private var workspaceId: String? = null
    private var scope: CacheScope? = null

    /** The queue as the cache last said, before search and channel. Null until it has said anything. */
    private var loaded: List<Conversation>? = null

    /** Whether this queue has been read from the server at least once in this scope. */
    private var everRead = false

    /** The last refresh's failure, for a queue with nothing cached to show instead. */
    private var failure: String? = null

    private var listJob: Job? = null
    private var countsJob: Job? = null

    /** Rows whose visitor intel has been asked for, so a re-emission does not ask again. */
    private val intelAsked = HashSet<String>()

    // MARK: - Loading

    fun bind(workspaceId: String) {
        if (this.workspaceId == workspaceId) return
        this.workspaceId = workspaceId
        scope = sync.scope(workspaceId)
        // A different workspace is a different plan and a different list, so
        // nothing from the old one carries over — including the channel
        // filter, which may name a plugin this workspace has never installed.
        _channel.value = null
        _query.value = ""
        _intel.value = emptyMap()
        intelAsked.clear()
        _counts.value = InboxCounts()
        _syncProblem.value = null
        observe()
        observeCounts()
        load()
        loadChannels()
    }

    /**
     * A queue and a channel are siblings in the menu, so they behave as
     * siblings here: choosing either one drops the other. This is iOS's rule
     * (`InboxViewModel.open`) and it is the only one that keeps the header
     * honest — a queue chosen while "Telegram" stayed underneath left the bar
     * naming a narrowing that the operator had just navigated away from, with
     * no obvious way back.
     */
    fun select(filter: InboxFilter) {
        val queueChanged = _filter.value != filter
        val channelDropped = _channel.value != null
        if (!queueChanged && !channelDropped) return
        _filter.value = filter
        _channel.value = null
        if (queueChanged) {
            // A different queue is a different question: carrying the search
            // terms across would silently filter a list nobody searched.
            _query.value = ""
            observe()
            observeCounts()
            load()
        } else {
            // Same queue, channel lifted: the list is already in hand.
            publish()
        }
    }

    fun selectChannel(key: String?) {
        if (_channel.value == key) return
        _channel.value = key
        // A channel inbox shows what is open on it, as on iOS. The queue has
        // to actually change for that to be true, and a queue change is a new
        // request — the conversations endpoint has no `channel` parameter, so
        // the channel itself is laid over whatever comes back.
        if (key != null && _filter.value != InboxFilter.OPEN) {
            _filter.value = InboxFilter.OPEN
            _query.value = ""
            observe()
            observeCounts()
            load()
        } else {
            publish()
        }
    }

    fun setQuery(value: String) {
        _query.value = value
        publish()
    }

    /** Pull-to-refresh: a real request, not a conditional one. */
    fun refresh() {
        _refreshing.value = true
        load(force = true)
    }

    /**
     * The cache's copy of this queue, for as long as this queue is the one on
     * screen. The first emission is whatever was saved — which on a warm
     * cache is the inbox, before any request has been made.
     */
    private fun observe() {
        val scope = scope ?: return
        val filter = _filter.value
        listJob?.cancel()
        loaded = null
        everRead = false
        failure = null
        publish()
        sync.coordinator.focusInbox(scope, filter)
        listJob = viewModelScope.launch {
            everRead = sync.conversations.hasList(scope, filter)
            sync.conversations.observeInbox(scope, filter).collect { rows ->
                loaded = rows
                publish()
                loadIntel(rows)
            }
        }
    }

    private fun observeCounts() {
        val scope = scope ?: return
        val queue = _filter.value.queue
        countsJob?.cancel()
        countsJob = viewModelScope.launch {
            sync.coordinator.counts.collect { snapshot ->
                if (snapshot != null && snapshot.matches(scope, queue)) _counts.value = snapshot.counts
            }
        }
    }

    private fun load(force: Boolean = false) {
        val scope = scope ?: return
        val filter = _filter.value
        viewModelScope.launch {
            val result = sync.coordinator.refreshInbox(scope, filter, force = force, reason = if (force) "pull" else "open")
            // The answer is for the queue that was asked about. If the
            // operator has moved on, the cache has it for when they return,
            // and this screen has nothing to say about it.
            if (this@InboxViewModel.scope == scope && _filter.value == filter) {
                result
                    .onSuccess {
                        everRead = true
                        failure = null
                        _syncProblem.value = null
                    }
                    .onFailure { error ->
                        val text = error.displayText(language())
                        failure = text
                        // Offline says what the operator is looking at: what
                        // this phone saved. Any other failure says itself.
                        _syncProblem.value = if (error.isOffline) StrAndroid.showingSaved(language()) else text
                    }
                publish()
            }
            _refreshing.value = false
            sync.coordinator.refreshCounts(scope, filter.queue)
        }
    }

    private fun loadChannels() {
        val workspace = workspaceId ?: return
        viewModelScope.launch {
            runCatching { api.channelInboxes(workspace) }.onSuccess { fresh ->
                if (workspaceId != workspace) return@onSuccess
                _channels.value = fresh
                // A channel that has just left the plan must not stay
                // selected with nothing behind it, or the list is filtered by
                // a name no row in the menu carries any more.
                if (_channel.value != null && fresh.none { it.key == _channel.value }) {
                    _channel.value = null
                    publish()
                }
            }
        }
    }

    /**
     * Where the visitors are and what they are on.
     *
     * One batched call for the rows it does not already know, rather than
     * one per row: fifty rows appearing at once would otherwise be fifty
     * requests, and a list that re-emits because one row changed would ask
     * for all fifty again. Kept in memory for the life of this workspace —
     * it is advisory decoration, not something worth a table.
     */
    private fun loadIntel(conversations: List<Conversation>) {
        val workspace = workspaceId ?: return
        val known = _intel.value.keys
        val ids = conversations.map { it.id }.filter { it !in known && it !in intelAsked }
        if (ids.isEmpty()) return
        intelAsked += ids
        viewModelScope.launch {
            runCatching { api.visitorIntelByConversation(workspace, ids) }
                .onSuccess { fresh -> if (workspaceId == workspace) _intel.update { it + fresh } }
                .onFailure { intelAsked -= ids.toSet() }
        }
    }

    // MARK: - What the screen sees

    private fun publish() {
        val rows = loaded
        _state.value = when {
            rows == null -> InboxState.Loading
            // Nothing cached and never read: the skeleton while the first
            // read is out, and its failure if it fails — the one case where
            // there is nothing better to show than the error.
            rows.isEmpty() && !everRead -> failure?.let(InboxState::Failed) ?: InboxState.Loading
            else -> InboxState.Loaded(narrow(rows))
        }
    }

    private fun narrow(rows: List<Conversation>): List<Conversation> {
        val terms = _query.value.trim()
        val channel = _channel.value
        return rows
            .filter { channel == null || it.channelKey == channel }
            .filter { terms.isEmpty() || it.matches(terms) }
    }

    /**
     * What counts as a match.
     *
     * Name, email, subject and the last message's body — the four things an
     * operator would be looking at when they reach for the magnifier. Case is
     * folded; the visitor code is included because an anonymous visitor has
     * nothing else to be searched by.
     */
    private fun Conversation.matches(terms: String): Boolean {
        val needle = terms.lowercase()
        return listOfNotNull(
            contact?.name,
            contact?.email,
            contact?.visitorCode,
            subject,
            Format.preview(lastMessage?.body),
        ).any { it.lowercase().contains(needle) }
    }

    /** The queues this plan (and the AI switches, for the AI queue) include. */
    fun filters(entitlements: Entitlements?, access: WorkspaceAccess, automated: Int?): List<InboxFilter> =
        InboxFilter.available(entitlements, access, automated)

    /** The subset that stays on the strip above the list. */
    fun chips(entitlements: Entitlements?, access: WorkspaceAccess, automated: Int?): List<InboxFilter> =
        InboxFilter.chips(entitlements, access, automated)
}
